### PR TITLE
feat: Deployment History Config Diff & Scope Variables Support in Comparison View across Config Diff

### PR DESCRIPTION
**Pull request by RohitRaj011**
# Description

Scope Variables Support in Config Diff

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [x] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] Does this PR require documentation updates?
- [ ] I've updated documentation as required by this PR.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx [hunk 1]
```diff
@@ -390,6 +465,8 @@ const getConfigMapSecretData = (
 
         return {
             id: `${resourceType === ConfigResourceType.ConfigMap ? EnvResourceType.ConfigMap : EnvResourceType.Secret}-${currentItem?.name || compareItem?.name}`,
+            pathType:
+                resourceType === ConfigResourceType.ConfigMap ? EnvResourceType.ConfigMap : EnvResourceType.Secret,
```
Comment (eshankvaish) [inline src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx:469]: Suggestion:

resourceType === ConfigResourceType.ConfigMap ? EnvResourceType.ConfigMap : EnvResourceType.Secret can be commoned out

File: src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx [hunk 2]
```diff
@@ -447,8 +549,36 @@ export const getAppEnvDeploymentConfigList = <ManifestView extends boolean = fal
...
+                currentDeploymentData.codeEditorValue.value,
+                compareDeploymentData.codeEditorValue.value,
+            ),
+        }
+
+        let currentPipelineConfigData: DeploymentHistoryDetail
+        let comparePipelineConfigData: DeploymentHistoryDetail
+        let pipelineConfigData: DeploymentConfigDiffProps['configList'][0]
+
+        if (_currentList.pipelineConfigData || _compareList.pipelineConfigData) {
```
Comment (eshankvaish) [inline src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx:562]: let's update the name for the current and compare list

File: src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx [hunk 3]
```diff
@@ -581,3 +741,31 @@ export const getDefaultVersionAndPreviousDeploymentOptions = (data: TemplateList
...
+
+export const diffStateTooltipTextMap: Record<DeploymentConfigDiffState, string> = {
+    hasDiff: 'File has difference',
+    added: 'File has been added',
+    deleted: 'File has been deleted',
+    noDiff: null,
+}
+
+export const diffStateTextColorMap: Record<DeploymentConfigDiffState, `c${string}`> = {
+    hasDiff: 'cy-7',
```
Comment (eshankvaish) [inline src/Shared/Components/DeploymentConfigDiff/DeploymentConfigDiff.utils.tsx:767]: Suggestion: move to constants

File: src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/utils.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+export const getPipelineDeploymentsWfrIds = ({
+    pipelineDeployments,
+    wfrId,
+}: { pipelineDeployments: History[] } & Pick<DeploymentHistoryConfigDiffProps, 'wfrId'>) => {
+    const currentDeploymentIndex = pipelineDeployments.findIndex(({ id }) => id === wfrId)
+    const previousWfrId = currentDeploymentIndex > -1 ? pipelineDeployments[currentDeploymentIndex + 1]?.id : null
+
+    return {
+        currentWfrId: wfrId,
+        previousWfrId: previousWfrId ?? null,
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/utils.ts:25]:
```suggestion
        previousWfrId,
```

File: src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiff.tsx [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    const [
+        deploymentTemplateResolvedDataLoader,
+        deploymentTemplateResolvedData,
+        deploymentTemplateResolvedDataErr,
+        reloadDeploymentTemplateResolvedData,
+    ] = useAsync(
+        () =>
+            abortPreviousRequests(
+                () =>
+                    Promise.all([
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiff.tsx:95]: suggestion: create a separate service with promise.all

File: src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiffCompare.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    renderRunSource,
+    resourceId,
+    ...props
+}: DeploymentHistoryDiffDetailedProps) => {
+    // HOOKS
+    const { path, params } = useRouteMatch()
+
+    // URL FILTERS
+    const { compareWfrId, updateSearchParams, sortBy, sortOrder, handleSorting } = useUrlFilters<
+        'sort-config' | '',
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiffCompare.tsx:35]: check if enum should be used. saw this at multiple places

**Review by eshankvaish — CHANGES_REQUESTED** [review]

File: src/Common/Helper.tsx [hunk 7]
```diff
@@ -937,3 +937,19 @@ export function asyncWrap(promise): any[] {
...
+
+    return (...args: Parameters<T>) => {
+        const now = Date.now()
+
+        if (now - lastCall >= delay) {
+            lastCall = now
+            func(...args)
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Common/Helper.tsx:955]: there's a useThrottleEffect hook  - different use-case

File: src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiffCompare.tsx (hunk 6 above)
Comment (RohitRaj011) [reply src/Shared/Components/CICDHistory/DeploymentHistoryConfigDiff/DeploymentHistoryConfigDiffCompare.tsx:35]: only one place

File: src/Shared/Helpers.tsx [hunk 8]
```diff
@@ -821,3 +821,34 @@ export const getDefaultValueFromType = (value: unknown) => {
...
+ * This function takes an array of objects and a key, and groups the objects in the array
+ * based on the value of the specified key. If an object does not have the specified key,
+ * it will be grouped under the `'UNGROUPED'` key.
+ *
+ * @param array - The array of objects to be grouped.
+ * @param key - The key of the object used to group the array.
+ * @returns An object where the keys are the unique values of the specified key in the array,
+ * and the values are arrays of objects that share the same key value.
+ */
+export const groupArrayByObjectKey = <T extends Record<string, any>, K extends keyof T>(
```
[the remark this answers is not in the document]
Comment (RohitRaj011) [reply src/Shared/Helpers.tsx:837]: different use-case can't be used

**Review by eshankvaish — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Review by eshankvaish — DISMISSED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**